src/include/txn.i

```diff
@@ -510,15 +510,24 @@ __wt_txn_pinned_timestamp(WT_SESSION_IMPL *session, wt_timestamp_t *pinned_tsp)
 
     /*
      * The read of the timestamp pinned by a checkpoint needs to be carefully ordered: if a
-     * checkpoint is starting and we have to use the checkpoint timestamp, we take the minimum of it
-     * with the oldest timestamp, which is what we want.
+     * checkpoint is starting and we have to use the oldest timestamp stored in the checkpoint, we
+     * take the minimum of it with the oldest timestamp, which is what we want.
+     */
+    WT_READ_BARRIER();
+    checkpoint_oldest_ts = txn_global->checkpoint_oldest_timestamp;
+    /*
+     * Make sure that we read the checkpoint's oldest timestamp first. Otherwise, if we are
+     * interrupted here, the checkpoint may have been completed and a new checkpoint has been
+     * started before we resume execution. In this case, we may find the oldest timestamp larger
+     * than the stable timestamp.
      */
     WT_READ_BARRIER();
-
     checkpoint_ts = txn_global->checkpoint_timestamp;
 
-    if (checkpoint_ts != 0 && checkpoint_ts < pinned_ts)
-        *pinned_tsp = checkpoint_ts;
+    if (checkpoint_ts != WT_TS_NONE && checkpoint_oldest_ts < pinned_ts) {
```

Comment:
The check of NONE is against `checkpoint_ts` and comparison is with `checkpoint_oldest_ts`. Later the `checkpoint_ts` is used only to assert. These conditions can lead to confusion. Why can't we calculate the Minimum of both `checkpoint_ts` and `checkpoint_oldest_ts` and compare it against pinned_tst?

Reply:
We only need to pin the timestamp if the checkpoint uses timestamp and that is determined by checkpoint_ts. It's possible checkpoint_oldest_ts is 0 and checkpoint_ts is not 0. If we take the minimum first, we will not pin timestamp 0 but actually we should.

Reply:
I understood your point, but 0 is not a valid timestamp, so I feel the scenario of having one value as 0 and another not possible.

Reply:
It's possible in the first checkpoint that oldest is still 0. We may have stable timestamp at 20 and oldest still at 0. If we only read oldest, we can't decide whether we need to pin it or not. Also we may start a checkpoint without a timestamp. ```
    /* Clear a checkpoint's pinned ID. */
    if (WT_SESSION_IS_CHECKPOINT(session)) {
        txn_global->checkpoint_txn_shared.pinned_id = WT_TXN_NONE;
        txn_global->checkpoint_timestamp = 0;
    }
```
Here we only clears checkpoint_timestamp not oldest_ts. May be we should do that as well. Anyway, in neither case, we can solely rely on the oldest timestamp itself to decide if we need to pin or not.

Reply:
ok. Thanks for explaining.